dimod/serialization/format.py

```diff
@@ -0,0 +1,143 @@
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# ================================================================================================
+from itertools import islice
+
+import numpy as np
+
+import dimod
+
+
+class PassVar(object):
+    def __str__(self):
+        return '..'
+
+
+class SampleSetFormatter(object):
+
+    min_column_width = 2
+    column_space = 2
+
+    max_rows = 40
+    max_columns = 12  # 17
+    header = True
+    footer = True
+    index = True
+
+    def __init__(self, sampleset):
+
+        self.sampleset = sampleset
+
+        # sampleset's size/shape/fields won't change, so we can build the (empty)
+        # array
+
+        nrows = min(self.max_rows,
+                    (len(sampleset) +  # one row for header
+                        self.header +  # two for footer
+                        2*self.footer))
+        nrows -= 2*self.footer  # two rows for footer
+
+        ncols = min(self.max_columns,
+                    (len(sampleset.variables) +
+                        len(sampleset.record.dtype.names) - 1 +
+                        self.index))
+
+        self.char = char = np.full((nrows, ncols), '', dtype=object)
+
+        # get columns
+        self.fields = fields = [field for field in sampleset.record.dtype.names
+                                if field != 'sample']
+
+        max_variables = self.max_columns - len(fields) - self.index
+        if len(sampleset.variables) > max_variables:
+            variables = sampleset.variables[:max_variables-2]
+            variables.append(PassVar())
+            variables.append(sampleset.variables[-1])
+        else:
+            # we can do all the variables
+            variables = sampleset.variables
+        variable_labels = list(map(str, variables))
+        self.variables = variables
+
+        if self.header:
+            char[0, :] = ['']*self.index + variable_labels + fields
+
+    def to_string(self):
+        sampleset = self.sampleset
+        num_samples = len(sampleset)
+        char = self.char
+        nrows, ncols = char.shape
+
+        variables = self.variables
+        fields = self.fields
+
+        if sampleset.vartype is dimod.SPIN:
+            def _str(datum, v):
+                if isinstance(v, PassVar):
+                    return str(v)
+                return '+1' if datum.sample[v] > 0 else '-1'
+        else:
+            def _str(datum, v):
+                if isinstance(v, PassVar):
+                    return str(v)
+                return '1' if datum.sample[v] > 0 else '0'
+
+        ci = 1*self.header
+        for i, datum in enumerate(sampleset.data()):
+            row = [str(i)]*self.header
+            row.extend(_str(datum, v) for v in variables)
+            row.extend(str(getattr(datum, field)) for field in fields)
+
+            char[ci, :] = row
+            ci += 1
+
+            if ci >= nrows:
```

Comment:
It is not obvious to me why you need `ci`. Why can't the `ci` in lines 103-106 be replaced with `i+1`? 

Maybe things would be clearer to me if I knew what `ci` stands for. I think it's "column index"; if so, it would be nice as a comment.